healvis/sky_model.py

```diff
@@ -168,3 +163,78 @@ def write_hdf5(self, filename):
                         dset = fileobj.create_dataset(k, data=d)
                     else:
                         dset = fileobj.create_dataset(k, data=d, compression='gzip', compression_opts=9)
+
+
+def flat_spectrum_noise_shell(sigma, freq_array, Nside, Nskies, ref_chan=0, shared_mem=False):
+    """
+    Make a flat-spectrum noise-like shell.
+
+    Args:
+        sigma : float
+            Power spectrum amplitude
+        freq_array : ndarray
+            Frequencies [Hz]
+        Nside : int
+            HEALpix Nside resolution
+        Nskies : int
+            Number of indepenent skies to simulate
+        ref_chan : int
+            freq_array reference channel index for comoving volume factor
+        shared_mem : bool
+            If True use mparray to generate data
+
+    Returns:
+        data : ndarray, shape (Nskies, Npix, Nfreqs)
+            EoR shell as HEALpix maps
+    """
+    # generate empty array
+    Nfreqs = len(freq_array)
+    Npix = hp.nside2npix(Nside)
+    if shared_mem:
+        data = mparray((Nskies, Npix, Nfreqs), dtype=float)
+    else:
+        data = np.zeros((Nskies, Npix, Nfreqs), dtype=float)
+
+    # setup parameters
+    dnu = np.diff(freq_array)[0]
+    om = 4 * np.pi / float(12 * Nside**2)
+    Zs = f21 / freq_array - 1.0
+    dV0 = comoving_voxel_volume(Zs[ref_chan], dnu, om)
+
+    # iterate over frequencies
+    for i in range(Nfreqs):
+        dV = comoving_voxel_volume(Zs[i], dnu, om)
+        amp = sigma * np.sqrt(dV0 / dV)
+        data[:, :, i] = np.random.normal(0.0, amp, (Nskies, Npix))
+
+    return data
+
+
+def gsm_shell(Nside, freq_array):
+    """
+    Generate a Global Sky Model shell
+
+    Args:
+        Nside : int
+            Nside resolution of HEALpix maps
+        freq_array : ndarray
+            Array of frequencies [Hz]
+
+    Returns:
+        data : ndarray, shape (Npix, Nfreqs)
+            GSM shell as HEALpix maps
+    """
+    assert pygsm_import, "Cannot use gsm_shell() when pygsm_import == False"
+
+    maps = pygsm.GlobalSkyModel(freq_unit='Hz', basemap='haslam').generate(freq_array)
+    # Units K
+    # The 2016 version is working, but a little harder to interpret.
```

Comment:
Should delete this comment.

It's a problem with the PyGSM module. The GSM2016 PCA components are missing from the repo, because the file is too large. There's an issue on the pygsm repo in which the author links to a google drive file with the full data, but trying to use GlobalSkyModel2016 will fail.

Also, even if it is working, the 2016 model is at Nside 1024, which slows down the coordinate rotation and ud_grade substantially. I doubt it will make much difference what resolution we start with given the Nside=128 ish resolutions we're simulating at.